rustls/src/hkdf.rs

```diff
@@ -0,0 +1,243 @@
+//! HKDF from RFC5869
+
+use crate::crypto::{hash, hmac};
+
+/// This is the inputs to HKDF-Extract, except for IKM.
+pub(crate) struct Extractor {
+    salt: Box<dyn hmac::Key>,
+    hmac: &'static dyn hmac::Hmac,
+}
+
+impl Extractor {
+    pub(crate) fn no_salt(hmac: &'static dyn hmac::Hmac) -> Self {
+        let zeroes = [0u8; hash::HASH_MAX_OUTPUT];
+        Self {
+            salt: hmac.open_key(&zeroes[..hmac.hash_output_len()]),
+            hmac,
+        }
+    }
+
+    pub(crate) fn with_salt(hmac: &'static dyn hmac::Hmac, salt: &[u8]) -> Self {
+        Self {
+            salt: hmac.open_key(salt),
+            hmac,
+        }
+    }
+
+    /// This is `HKDF-Extract`.
+    pub(crate) fn extract(self, ikm: &[u8]) -> Expander {
+        Expander(
+            self.hmac
+                .open_key(self.salt.sign(&[ikm]).as_ref()),
+        )
+    }
+}
+
+pub(crate) struct OutputLengthError;
+
+/// This is a PRK ready for use via `expand()` et al.
+pub(crate) struct Expander(Box<dyn hmac::Key>);
+
+/// This is a single block output from HKDF-Expand.
+#[derive(Clone)]
+pub(crate) struct OkmOneBlock(hmac::Tag);
+
+impl AsRef<[u8]> for OkmOneBlock {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl Expander {
+    pub(crate) fn from_okm(okm: &OkmOneBlock, hmac: &'static dyn hmac::Hmac) -> Self {
+        Self(hmac.open_key(okm.0.as_ref()))
+    }
+
+    fn expand_unchecked(&self, info: &[&[u8]], output: &mut [u8]) {
```

Comment:
I would order this as `expand_slice()`, `expand_one_block()`, `expand()`, `expand_unchecked()`, `one_block_len()`.